### ISSUE TITLE
Problem with compareECL tests in opm-simulators.

### ISSUE THREAD

**Issue by dr-robertk** [OPM/opm-simulators#1192]
On the current and probably also older master I have the following problem. 
Some of the SPE cases don't work, even if they're supposed to. I checked that my opm-data repo is up2date ans so if every other opm module.

      Start 17: compareECLFiles_flow_ebos+SPE1CASE2
17/37 Test #17: compareECLFiles_flow_ebos+SPE1CASE2 ................***Failed   18.68 sec
      Start 18: compareECLFiles_flow_ebos+SPE1CASE2_2P
18/37 Test #18: compareECLFiles_flow_ebos+SPE1CASE2_2P .............***Failed    1.41 sec
      Start 19: compareECLFiles_flow_legacy+SPE1CASE2
19/37 Test #19: compareECLFiles_flow_legacy+SPE1CASE2 ..............***Failed   17.83 sec
      Start 20: compareECLFiles_flow_legacy+SPE1CASE2_2P
20/37 Test #20: compareECLFiles_flow_legacy+SPE1CASE2_2P ...........***Failed    2.53 sec
      Start 21: compareECLFiles_flow_sequential+SPE1CASE1
21/37 Test #21: compareECLFiles_flow_sequential+SPE1CASE1 ..........   Passed    2.88 sec
      Start 22: compareECLFiles_flow_ebos+SPE3CASE1
22/37 Test #22: compareECLFiles_flow_ebos+SPE3CASE1 ................***Failed    5.53 sec
      Start 23: compareECLFiles_flow_legacy+SPE3CASE1
23/37 Test #23: compareECLFiles_flow_legacy+SPE3CASE1 ..............***Failed   13.60 sec
      Start 24: compareECLFiles_flow_ebos+SPE9_CP_SHORT
24/37 Test #24: compareECLFiles_flow_ebos+SPE9_CP_SHORT ............   Passed    3.82 sec
      Start 25: compareECLFiles_flow_legacy+SPE9_CP_SHORT
25/37 Test #25: compareECLFiles_flow_legacy+SPE9_CP_SHORT ..........   Passed    8.37 sec
      Start 26: compareECLFiles_flow_multisegment+2D_H__
26/37 Test #26: compareECLFiles_flow_multisegment+2D_H__ ...........   Passed    3.90 sec
      Start 27: compareRestartedSim_flow_legacy+SPE1CASE2_ACTNUM
27/37 Test #27: compareRestartedSim_flow_legacy+SPE1CASE2_ACTNUM ...   Passed    0.87 sec
      Start 28: compareRestartedSim_flow_legacy+SPE9_CP_SHORT
28/37 Test #28: compareRestartedSim_flow_legacy+SPE9_CP_SHORT ......   Passed   10.68 sec
      Start 29: compareRestartedSim_flow_ebos+SPE1CASE2_ACTNUM
29/37 Test #29: compareRestartedSim_flow_ebos+SPE1CASE2_ACTNUM .....   Passed    0.61 sec
      Start 30: compareRestartedSim_flow_ebos+SPE9_CP_SHORT
30/37 Test #30: compareRestartedSim_flow_ebos+SPE9_CP_SHORT ........   Passed    5.16 sec
      Start 31: compareECLInitFiles_flow_legacy+NORNE_ATW2013
31/37 Test #31: compareECLInitFiles_flow_legacy+NORNE_ATW2013 ......   Passed    3.09 sec
      Start 32: compareECLInitFiles_flow_ebos+NORNE_ATW2013
32/37 Test #32: compareECLInitFiles_flow_ebos+NORNE_ATW2013 ........   Passed    3.39 sec
      Start 33: compareParallelSim_flow_mpi+SPE1CASE2
33/37 Test #33: compareParallelSim_flow_mpi+SPE1CASE2 ..............   Passed    5.15 sec
      Start 34: compareParallelSim_flow_mpi+SPE9_CP_SHORT
34/37 Test #34: compareParallelSim_flow_mpi+SPE9_CP_SHORT ..........   Passed   22.70 sec
      Start 35: compareParallelSim_flow_ebos+SPE1CASE2
35/37 Test #35: compareParallelSim_flow_ebos+SPE1CASE2 .............   Passed    2.47 sec
      Start 36: compareParallelSim_flow_ebos+SPE9_CP_SHORT
36/37 Test #36: compareParallelSim_flow_ebos+SPE9_CP_SHORT .........   Passed   14.39 sec
      Start 37: compareParallelSim_flow_mpi+SPE3CASE1
37/37 Test #37: compareParallelSim_flow_mpi+SPE3CASE1 ..............   Passed    4.62 sec

84% tests passed, 6 tests failed out of 37

Total Test time (real) = 156.80 sec

The following tests FAILED:
         17 - compareECLFiles_flow_ebos+SPE1CASE2 (Failed)
         18 - compareECLFiles_flow_ebos+SPE1CASE2_2P (Failed)
         19 - compareECLFiles_flow_legacy+SPE1CASE2 (Failed)
         20 - compareECLFiles_flow_legacy+SPE1CASE2_2P (Failed)
         22 - compareECLFiles_flow_ebos+SPE3CASE1 (Failed)
         23 - compareECLFiles_flow_legacy+SPE3CASE1 (Failed)
Errors while running CTest

**Comment by akva2** [comment]
ctest -V please

**Comment by andlaus** [comment]
this looks similar to the situation on my machine if I use too aggressive optimization flags. does this disappear when using plain `-O2`?

**Comment by dr-robertk** [comment]
[ctest-out.txt](https://github.com/OPM/opm-simulators/files/1035516/ctest-out.txt)

**Comment by dr-robertk** [comment]
This should not happen with -O3, otherwise I would consider this a bug.

**Comment by andlaus** [comment]
yes, I agree that this would constitute be a minor bug, probably because of the tolerances used for the tests. `-march=native`, `-ffast-math` and the tree-vectorization flags also affect the results slightly...

**Comment by dr-robertk** [comment]
Same problems occur with less optimized executable "-fPIC -O2 -DNDEBUG".

**Comment by joakim-hove** [comment]
I assume we are now comparing simulation results from @dr-robertk's computer with the results generated on the build server - is it even realistic to assume equailty here as long as the only common ingredient is "some recently new linux distribution"?

**Comment by dr-robertk** [comment]
@joakim-hove: That is a good question. I would say it is, otherwise it should not be in the selection of "test". I guess the common development process is to run ctest to see if everything works as expected before creating a PR.

**Comment by dr-robertk** [comment]
Maybe I should rephrase my question. How can I check what is going wrong in #1186  when the test fail even for the master?

**Comment by akva2** [comment]
ctest -V, possibly combined with hacking test/run-regressionTest.sh and tests/analyze_summary.failure.sh or tests/analyze_ecl_failure.sh. 

while i've been silent i am working on getting a data set large enough so i can possibly adjust tolerances with some confidence. but i didn't find enough time to build the whole set yesterday, and i'm traveling tuesdays so i had to put it off until tomorrow or so. what i can say is that the differences are huge across platforms and toolchains and i am not convinced this is not a bug yet.

**Comment by andlaus** [comment]
> what i can say is that the differences are huge across platforms and toolchains and i am not convinced this is not a bug yet.

I think that if it turns out that the differences due to the tooling make the unit tests fail, the tolerances for the unit tests should be considered to be too tight to make these tests useful to be run during the development process. possibly, the default tolerances of the build system should thus be relaxed so that they work for everyone while the jenkins rig still sticks to tighter ones so that it continues to catch even minor changes?!

**Comment by akva2** [comment]
we'll see once i have *real information* and not guesses at hand.

**Comment by atgeirr** [comment]
> I think that if it turns out that the differences due to the tooling make the unit tests fail, the tolerances for the unit tests should be considered to be too tight to make these tests useful to be run during the development process. possibly, the default tolerances of the build system should thus be relaxed so that they work for everyone while the jenkins rig still sticks to tighter ones so that it continues to catch even minor changes?!

I think that would just cause endless headaches... For me up to now at least, usually there is consistency between failing locally and on Jenkins. Losing that means you have to go back to the code due to Jenkins failing a lot more often, which means a longer feedback loop.

**Comment by akva2** [comment]
i didn't dig very deep, hard without my tools (ie my desktop), but https://pastebin.com/HdHTHFBP  is definitely not helping.

**Comment by dr-robertk** [comment]
I would rather suggest to tighten the solver tolerances for the small SPE cases might be a good idea to avoid epsilon errors.

**Comment by akva2** [comment]
yeah. however i did initially push these as low as the solver setup at the time allowed without causing time step cuts (which makes things a lot less predictable) but hopefully there is some room left now since quite a lot has changed. but yes, that is my line of thought as well, we need stricter control over the obtained solution, not less control through allowing larger errors.

**Comment by akva2** [comment]
@andlaus note the pastebin.

**Comment by andlaus** [comment]
Ok, just did this: that valgrind complaint seems to be caused somewhere in the well model. @totto82: can you have a quick look?

**Comment by atgeirr** [comment]
@akva2 which case/arguments did you use to produce the pastebin report? Also, should I interpret the memcheck reports to mean there are only two invalid read events (i.e. not one per timestep or iteration) during the complete run?

**Comment by akva2** [comment]
The case and parametets listed at the very top of the log....

**Comment by dr-robertk** [comment]
I have checked the test with improved accuracy for the Newton solvers with the result that now all tests fail. For me this means: The tests cannot be trusted since they only tell when a result is not within bounds of a previously computed (possibly wrong) solution. This has to be improved.

**Comment by alfbr** [comment]
> The tests cannot be trusted since they only tell when a result is not within bounds of a previously computed (possibly wrong) solution.

Yes, this is correct. However, there is no such thing as the gold standard solution, so this is the best we can do at the current stage. We will probably never get a perfect solution here, but you are certainly welcome to help improve the current state.

**Comment by dr-robertk** [comment]
@alfbr golden standard or not, do you really want to call "tests giving false alarm on just about every change that is somewhat related to performance implementations for better or worse" the best we can do?

**Comment by joakim-hove** [comment]
> golden standard or not, do you really want to call "tests giving false alarm on just about every change that is somewhat related to performance implementations for better or worse" the best we can do?

>Yes, this is correct. However, there is no such thing as the gold standard solution, so this is the best we can do at the current stage. We will probably never get a perfect solution here, but you are certainly welcome to help improve the current state.

OK - I'll bite: In my opinion the problem is that we have far too few unit tests. Ideally the various repositories should have enough unit tests that a developer could be 99% certain everything was correct when making the PR. Right now way too much of the testing is delegated all the way down to a fully integrated flow simulation,and when that fails all you is that "something failed".

**Comment by alfbr** [comment]
> do you really want to call "tests giving false alarm on just about every change that is somewhat related to performance implementations for better or worse" the best we can do?

No, of course not. What I meant was that comparing to our own results is the only option for having strict bounds without frequently triggering false alarms. We have no "correct" answer to compare with, and attempting to use Eclipse results for that is bound to make the tests less valuable. I may have misunderstood what you meant though. Finding a good balance on how strict bounds the tests should have should be considered an ongoing task open for discussion.

**Comment by dr-robertk** [comment]
@joakim-hove  is right with the unit tests, however, that is usually difficult for simulations codes. 

The problem that I see with the current SPE tests is, that they also fail if the solution gets "better", for example. If you make the solver tolerance smaller, i..e. the solver more accurate, then the tests also fail. So my question was: Is that really what we want? Along that lines I think the tests should work on every somewhat up2date linux/unix distribution. 

Then my ultimate goal was to figure out how the PRs related to performance can get into the master. Because right now I don't have the slightest idea what to fix or change, so that SPE3CASE1 or whatever silly test can run and the travis and jenkins are happy.

**Comment by alfbr** [comment]
> The problem that I see with the current SPE tests is, that they also fail if the solution gets "better",

Yes, and in this case it seems we have good cause to increase tolerances. I am however very glad we did not just increase them on the first cross-roads, these tests have already uncovered bugs a couple of times when the first suggestion was simply increasing tolerances.

**Comment by GitPaean** [comment]
Does this issue still apply?
